server/athenian/api/align/mutations/goals.py

```diff
@@ -0,0 +1,129 @@
+
+    if len({team_goal.team_id for team_goal in team_goals}) < len(team_goals):
+        raise CreateGoalError("More than one team goal with the same teamId")
+
+    valid_from, expires_at = _convert_goal_dates(input["validFrom"], input["expiresAt"])
+
+    goal = Goal(
+        account_id=account_id,
+        template_id=template_id,
+        valid_from=valid_from,
+        expires_at=expires_at,
+    )
+    return _CreateGoalInputParseResult(goal, team_goals)
+
+
+def _parse_team_goal_input(team_goal_input: dict) -> TeamGoal:
+    """Parse TeamGoalInput into a Team model."""
+    try:
+        # get the first non null value in GoalTargetInput
+        target = next(tgt for tgt in team_goal_input["target"].values() if tgt is not None)
+    except StopIteration:
+        raise CreateGoalError(f'Invalid target for teamId {team_goal_input["teamId"]}')
+
+    return TeamGoal(team_id=team_goal_input["teamId"], target=target)
+
+
+def _convert_goal_dates(valid_from: date, expires_at: date) -> Tuple[datetime, datetime]:
+    """Convert date objects from API into datetimes."""
+    valid_from = datetime.combine(valid_from, time.min, tzinfo=timezone.utc)
+    # expiresAt semantic is to include the given day, so datetime is set to the start of the
+    # following day
+    expires_at = datetime.combine(expires_at + timedelta(days=1), time.min, tzinfo=timezone.utc)
+    return (valid_from, expires_at)
+
+
+async def _parse_result_db_validation(
+    parse_result: _CreateGoalInputParseResult, sdb_conn: Connection,
+) -> None:
+    """Execute validation on parse result using the DB."""
+    # check that all team exist and belong to the right account
+    team_ids = [team_goal.team_id for team_goal in parse_result.team_goals]
+    teams_stmt = sa.select([Team.id]).where(
+        sa.and_(Team.id.in_(team_ids), Team.owner_id == parse_result.goal.account_id),
+    )
+    existing_team_ids_rows = await sdb_conn.fetch_all(teams_stmt)
+    existing_team_ids = sorted(r[0] for r in existing_team_ids_rows)
+
+    if sorted(team_ids) != existing_team_ids:
+        missing = [team_id for team_id in team_ids if team_id not in existing_team_ids]
+        missing_repr = ",".join(str(team_id) for team_id in missing)
+        raise CreateGoalError(f"Invalid teamId: {missing_repr}")
+
+
+async def _insert_db_goal(parse_result: _CreateGoalInputParseResult, sdb_conn: Connection) -> int:
+    """Insert the goal and related objects into DB."""
+    goal_values = parse_result.goal.create_defaults().explode()
+    new_goal_id = await sdb_conn.execute(sa.insert(Goal).values(goal_values))
+    team_goals_values = [
+        {
+            **team_goal.create_defaults().explode(with_primary_keys=True),
+            # goal_id can only be set now that Goal has been inserted
+            "goal_id": new_goal_id,
+        }
+        for team_goal in parse_result.team_goals
+    ]
+    await sdb_conn.execute(sa.insert(TeamGoal).values(team_goals_values))
```

Comment:
`sdb_conn.execute_many` is much more efficient, we should use it everywhere to insert several records.

Reply:
done